### PR TITLE
Add type to MooseApp

### PR DESCRIPTION
**Pull request by permcody**
Adding "type" to MooseApp. We may add this to the standard header in the future.

Note: Also renamed obj_name -> app_type in the AppFactory.

### CODE REVIEW

**Comment by friedmud** [comment]
I will go on record as saying I'm not the biggest fan of this kind of thing.  It's anti-polymorphism...

**Comment by permcody** [comment]
Generally speaking I would agree. However as long as it's not abused it's not exactly terrible. Yaqi's use for comparing matching types is not well handled through polymorphism (because most people won't bother to override the method unless it's abstract but then that's heavy handed) and using RTTI is always nasty business. There are certainly other ways of handling this without using a stored type like this. However I was also thinking about displaying the app type in a Multiapp run as the headers are printed out for each app.

File: framework/include/base/MooseApp.h [hunk 1]
```diff
@@ -66,6 +66,12 @@ class MooseApp : public libMesh::ParallelObject
    */
   InputParameters & parameters() { return _pars; }
 
+  /**
+   * Get the type of this obejct as a string
```
Comment (jwpeterson) [inline framework/include/base/MooseApp.h]: "obejct"

**Comment by jwpeterson** [comment]
:+1: I think we all agree that this is anti-polymorphism, but still useful.  Does it still need more discussion?

**Comment by jwpeterson** [comment]
@permcody is going to tweak this PR a bit more...

**Comment by permcody** [comment]
@friedmud - There is actually a valid use for this member outside of whatever plans Yaqi has for it. We chatted the other day about a "run_as_app" idea. This is the select box idea where you choose which application your staticly-linked executable behaves as during run time.

e.g. If I have "mammoth" compiled but I want to run a "rattlesnake" input file exactly as it would run in a rattlesnake binary this `type()` method should return "RattleSnakeApp" for that case.

Perhaps `type()` isn't the right name here?

**Comment by permcody** [comment]
After further thought, I believe this PR can be merged. I'm not going to build support for the "run_as" idea right now since it is non-trivial to accomplish _before_ the application is constructed. A static method would need to be created and jump through several hoops to return the registered names and a valid selection from an input file. I've capture this in a new ticket #5063 so this PR can be merged and closed.

**Pull request merged**